llama-index-integrations/llms/llama-index-llms-anthropic/llama_index/llms/anthropic/utils.py

```diff
@@ -40,32 +40,10 @@ def messages_to_anthropic_messages(
     return anthropic_messages, system_prompt
 
 
-# Function used in bedrock
-def _message_to_anthropic_prompt(message: ChatMessage) -> str:
-    if message.role == MessageRole.USER:
-        prompt = f"{HUMAN_PREFIX} {message.content}"
-    elif message.role == MessageRole.ASSISTANT:
-        prompt = f"{ASSISTANT_PREFIX} {message.content}"
-    elif message.role == MessageRole.SYSTEM:
-        prompt = f"{message.content}"
-    elif message.role == MessageRole.FUNCTION:
-        raise ValueError(f"Message role {MessageRole.FUNCTION} is not supported.")
-    else:
-        raise ValueError(f"Unknown message role: {message.role}")
-
-    return prompt
-
-
-def messages_to_anthropic_prompt(messages: Sequence[ChatMessage]) -> str:
+def messages_to_anthropic_prompt(messages: Sequence[ChatMessage]) -> Sequence[Dict]:
     if len(messages) == 0:
         raise ValueError("Got empty list of messages.")
 
-    # NOTE: make sure the prompt ends with the assistant prefix
-    if messages[-1].role != MessageRole.ASSISTANT:
-        messages = [
-            *list(messages),
-            ChatMessage(role=MessageRole.ASSISTANT, content=""),
-        ]
-
-    str_list = [_message_to_anthropic_prompt(message) for message in messages]
-    return "".join(str_list)
+    return [
+        {"role": message.role.value, "content": message.content} for message in messages
```

Comment:
I know anthropic requests alternating user/assistant prompts. We might have to check for that and merge neighbouring messages? The `llama-index-llms-anthropic` package does something similar

We can also pull out the system prompt here and pass it in under the `system` kwarg (assuming bedrock supports that same kwarg like the normal anthropic api)

Reply:
I'll make that update, but they will only work with [Claude 2.1 and beyond](https://docs.anthropic.com/claude/docs/system-prompts#where-can-i-use-system-prompts).

Reply:
> I know anthropic requests alternating user/assistant prompts. We might have to check for that and merge neighbouring messages? The `llama-index-llms-anthropic` package does something similar

I see where the non-Bedrock Anthropic logic extracts the system prompt [here](https://github.com/run-llama/llama_index/blob/0ae69d46e3735a740214c22a5f72e05d46d92635/llama-index-integrations/llms/llama-index-llms-anthropic/llama_index/llms/anthropic/utils.py#L35), but don't see any sign of logic that checks "neighbouring messages".

Reply:
I do see where the stream `gen` method differs [here](https://github.com/run-llama/llama_index/blob/756457898dcf474752c35f1ff580596dd9a2558f/llama-index-integrations/llms/llama-index-llms-bedrock/llama_index/llms/bedrock/base.py#L255) for Bedrock, but it uses a different class: `CompletionResponseGen`, which relies on the streaming text.  Anthropic meanwhile uses `ChatResponse` [here](https://github.com/run-llama/llama_index/blob/0ae69d46e3735a740214c22a5f72e05d46d92635/llama-index-integrations/llms/llama-index-llms-anthropic/llama_index/llms/anthropic/base.py#L192), which does support the `Message` object with a role.  Since the other Bedrock models share the same logic, deviating from that would require adding a new method at the `Bedrock` level for Anthropic alone.

Reply:
> but system prompts will only work with [Claude 2.1 and beyond](https://docs.anthropic.com/claude/docs/system-prompts#where-can-i-use-system-prompts).

Maybe I am doing something wrong but as per my observations system prompts don't seem to work in any claude model when passed as chat.

```python
llm = Bedrock(
    model="anthropic.claude-3-sonnet-20240229-v1:0", # same with 2.1
    region_name='us-east-1'
)

messages2 = [
    ChatMessage(
        role=MessageRole.SYSTEM, content="You are a joker."
    ),
    ChatMessage(
        role=MessageRole.USER, content="Tell me a joke."
    ),
]

llm.chat(messages2)
```

Reply:
> > but system prompts will only work with [Claude 2.1 and beyond](https://docs.anthropic.com/claude/docs/system-prompts#where-can-i-use-system-prompts).
> 
> Maybe I am doing something wrong but as per my observations system prompts don't seem to work in any claude model when passed as chat.
> 
> ```python
> llm = Bedrock(
>     model="anthropic.claude-3-sonnet-20240229-v1:0", # same with 2.1
>     region_name='us-east-1'
> )
> 
> messages2 = [
>     ChatMessage(
>         role=MessageRole.SYSTEM, content="You are a joker."
>     ),
>     ChatMessage(
>         role=MessageRole.USER, content="Tell me a joke."
>     ),
> ]
> 
> llm.chat(messages2)
> ```

I think that is right (based on my reading of the implementation).  They don't seem to be used in the `Bedrock` class which invokes the Anthropic provider.

Reply:
Perhaps you can merge down the message from system/ others to neighborhood user, or create a new user space. Not sure how effective the results would be but the API should be stable.